pandas/core/indexers.py

```diff
@@ -133,34 +133,48 @@ def check_setitem_lengths(indexer, value, values) -> None:
 
     Returns
     -------
-    None
+    bool
+        Whether this is an empty listlike setting which is a no-op.
 
     Raises
     ------
     ValueError
         When the indexer is an ndarray or list and the lengths don't match.
     """
-    # boolean with truth values == len of the value is ok too
+    no_op = False
+
     if isinstance(indexer, (np.ndarray, list)):
-        if is_list_like(value) and len(indexer) != len(value):
-            if not (
-                isinstance(indexer, np.ndarray)
-                and indexer.dtype == np.bool_
-                and len(indexer[indexer]) == len(value)
-            ):
-                raise ValueError(
-                    "cannot set using a list-like indexer "
-                    "with a different length than the value"
-                )
+        # We can ignore other listlikes becasue they are either
+        #  a) not necessarily 1-D indexers, e.g. tuple
+        #  b) boolean indexers e.g. BoolArray
+        if is_list_like(value):
+            if len(indexer) != len(value):
+                if not (
+                    isinstance(indexer, np.ndarray)
+                    and indexer.dtype == np.bool_
+                    and len(indexer[indexer]) == len(value)
+                ):
+                    # boolean with truth values == len of the value is ok too
+                    raise ValueError(
+                        "cannot set using a list-like indexer "
+                        "with a different length than the value"
+                    )
+            if len(indexer) == 0:
+                no_op = True
 
     elif isinstance(indexer, slice):
         # slice
-        if is_list_like(value) and len(values):
-            if len(value) != length_of_indexer(indexer, values):
-                raise ValueError(
-                    "cannot set using a slice indexer with a "
-                    "different length than the value"
-                )
+        if is_list_like(value):
+            if len(values):
+                if len(value) != length_of_indexer(indexer, values):
+                    raise ValueError(
+                        "cannot set using a slice indexer with a "
+                        "different length than the value"
+                    )
+            else:
+                # TODO: dont we still need lengths to match?
```

Comment:
I think yes? 
But can you do that here, since the original code (I think?) did it.

Reply:
the existing code does not check for length-match in the len==0 case.  ill take a look at adding that check and getting rid of this comment

Reply:
It actually is checked right now, because of the `len(key) != len(value)` before checking `elif not len(key): ` (but it's clearly not tested ..)

As example:
```
In [1]: arr = pd.date_range("2012", periods=3)._data                                                                                                                                                               

In [2]: arr                                                                                                                                                                                                        
Out[2]: 
<DatetimeArray>
['2012-01-01 00:00:00', '2012-01-02 00:00:00', '2012-01-03 00:00:00']
Length: 3, dtype: datetime64[ns]

In [3]: arr[[]] = [pd.Timestamp("2012")]                                                                                                                                                                           
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-d4b75e105bf5> in <module>
----> 1 arr[[]] = [pd.Timestamp("2012")]

~/scipy/pandas/pandas/core/arrays/datetimelike.py in __setitem__(self, key, value)
    620                         f"'{len(value)}'."
    621                     )
--> 622                     raise ValueError(msg)
    623                 elif not len(key):
    624                     return

ValueError: shape mismatch: value array of length '0' does not match indexing result of length '1'.
```

Reply:
Good catch; i was comparing against the existing `check_setitem_lengths`